src/libAtomVM/bif.c

```diff
@@ -1485,51 +1508,109 @@ term bif_erlang_bxor_2(Context *ctx, uint32_t fail_label, int live, term arg1, t
+{
+    if (lshift > 64) {
+        *out = 0;
+        return (n != 0);
     }
 
-    int64_t a = term_maybe_unbox_int64(arg1);
-    avm_int_t b = term_to_int(arg2);
-    int64_t result = op(a, b);
-
-    #if BOXED_TERMS_REQUIRED_FOR_INT64 > 1
-        return make_maybe_boxed_int64(ctx, fail_label, live, result);
-    #else
-        return make_maybe_boxed_int(ctx, fail_label, live, result);
-    #endif
+    int64_t res = (int64_t) (((uint64_t) n) << lshift);
+    *out = res;
+    int64_t check = int64_bsr(res, lshift);
+    return check != n;
 }
 
-static inline int64_t bsl(int64_t a, avm_int_t b)
+static inline int64_t int64_bsr_safe(int64_t n, unsigned int rshift)
 {
-    // TODO check for overflow
-    return a << b;
+    if (rshift > 64) {
+        return n < 0 ? -1 : 0;
+    }
+    return int64_bsr(n, rshift);
 }
 
 term bif_erlang_bsl_2(Context *ctx, uint32_t fail_label, int live, term arg1, term arg2)
 {
-    return bitshift_helper(ctx, fail_label, live, arg1, arg2, bsl);
-}
+    if (LIKELY(term_is_any_integer(arg1) && term_is_any_integer(arg2))) {
+        size_t arg1_size = term_is_integer(arg1) ? 0 : term_boxed_size(arg1);
+        avm_int_t b = term_to_int(arg2);
+        if (arg1_size <= BOXED_TERMS_REQUIRED_FOR_INT64) {
+            int64_t a = term_maybe_unbox_int64(arg1);
+            int64_t result;
+            if (!int64_bsl_overflow(a, b, &result)) {
+                #if BOXED_TERMS_REQUIRED_FOR_INT64 > 1
+                    return make_maybe_boxed_int64(ctx, fail_label, live, result);
+                #else
+                    return make_maybe_boxed_int(ctx, fail_label, live, result);
+                #endif
+            }
+        }
 
-static inline int64_t bsr(int64_t a, avm_int_t b)
-{
-    // TODO check for underflow
-    return a >> b;
+        intn_digit_t tmp_buf1[INTN_INT64_LEN];
+        intn_digit_t tmp_buf2[INTN_INT64_LEN];
+        intn_digit_t *m;
+        size_t m_len;
+        intn_integer_sign_t m_sign;
+        intn_digit_t *n;
+        size_t n_len;
+        intn_integer_sign_t n_sign;
+        args_to_bigint(arg1, arg2, tmp_buf1, tmp_buf2, &m, &m_len, &m_sign, &n, &n_len, &n_sign);
+
+        intn_digit_t bigres[INTN_MAX_RES_LEN];
+        size_t bigres_len = intn_bsl(m, m_len, b, bigres);
+
+        return make_bigint(ctx, fail_label, live, bigres, bigres_len, m_sign);
+
+    } else {
+        RAISE_ERROR_BIF(fail_label, BADARITH_ATOM);
+    }
 }
 
 term bif_erlang_bsr_2(Context *ctx, uint32_t fail_label, int live, term arg1, term arg2)
 {
-    return bitshift_helper(ctx, fail_label, live, arg1, arg2, bsr);
+    if (LIKELY(term_is_any_integer(arg1) && term_is_any_integer(arg2))) {
+        size_t arg1_size = term_is_integer(arg1) ? 0 : term_boxed_size(arg1);
+
+        avm_int_t b = term_to_int(arg2);
+
+        if (arg1_size <= BOXED_TERMS_REQUIRED_FOR_INT64) {
+            uint64_t a = (uint64_t) term_maybe_unbox_int64(arg1);
+            int64_t result = int64_bsr_safe(a, b);
```

Comment:
int64_bsr_safe takes `int64_t` so you don't need to cast the result of `term_maybe_unbox_int64` to `uint64_t`

Reply:
Fixed.